### PR TITLE
Proper glob post-filtering

### PR DESCRIPTION
**Pull request by dpanth3r**
- Fixes memory driver behaviour
- Cleaned tests up

---
This change is [Reviewable](https://reviewable.io/reviews/criteo/biggraphite/155)

### CODE REVIEW

**Comment by dpanth3r** [comment]
Among other things, closes #154

**Comment by dpanth3r** [comment]
For potential reviewer questions: I moved some features from graphite_utils to utils because it kind of made sense to me, AND it allowed to use the glob helpers from the memory driver (graphite_utils used to include utils, which would create an import cycle).

**Comment by iksaif** [comment]
So, that was my version of the same thing: https://github.com/iksaif/biggraphite/commit/8028bd9677ff30478e7f534681d2c888e0efaf46

graphite-utils and utils where separated on purpose because we didn't want too many graphite specific things to leak into the common part (to make it easier to have opentsdb/influxdb compatibility layer later).

The idea was that the accessor would support only wildcards, that the anything smarter would be done somewhere else.

**Comment by dpanth3r** [comment]
I would tend to say your suggested implementation is less understandable, and I believe mine will probably be a bit more efficient because it only uses a single regexp instead of as many as there are variants (which in the end makes little sense, apart maybe from the fact that it hides that between the fnmatch API).

I don't really understand your last point, because I did not move the smart stuff anywhere else. Unless you mean that about the in-memory driver, in which case I did it because it was already doing semi-smart things.

**Comment by dpanth3r** [comment]
There is now a new more correct version of the glob_to_regex feature, improving on fnmatch. I have also moved the glob-related code outside of graphite_utils, and updated the pre-filtering code of memory in the process to be glob/fnmatch agnostic (well, ok, it still tries to detect whether there is a globbing expression in components, but it has to!). This latest version of the patch includes additional tests.

File: biggraphite/drivers/memory.py [hunk 1]
```diff
@@ -118,14 +116,26 @@ def create_metric(self, metric):
...
-        glob_re = re.compile(fnmatch.translate(glob))
-
-        return [
-            name for name in names
-            if name.count(".") == dots_count and glob_re.match(name)
+        dots = glob.count(".")
+        literal_glob_components = [
+            (index, component)
+            for (index, component) in enumerate(glob.split("."))
+            if not bg_glob._is_graphite_glob(component)
```
Comment (iksaif) [inline biggraphite/drivers/memory.py]: Can we avoid *graphite* specific thing in the drivers ?
In this case:
* basic wildcard globbing should be done in the driver
* advanced integration specific globbing (graphite) should be callbed by the driver

Suggestion: add a specific method for wildcard globbing in bg_glob that could be callbed by drivers who can't do wildcards themselves

File: biggraphite/utils.py [hunk 2]
```diff
@@ -48,11 +48,11 @@ class ConfigError(Error):
 def accessor_from_settings(settings):
     """Get Accessor from configuration.
 
-    Args:
```
Comment (iksaif) [inline biggraphite/utils.py]: The convention we use is really "Args"

Comment (dpanth3r) [inline biggraphite/utils.py]: heh.

File: biggraphite/drivers/memory.py (hunk 1 above)
Comment (dpanth3r) [inline biggraphite/drivers/memory.py]: I'll add a helper in glob utils then, I knew this would come up :-)

Comment (dpanth3r) [inline biggraphite/drivers/memory.py]: Done.

File: biggraphite/glob_utils.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+      A tuple:
+        First element: sorted list of Cassandra metrics matched by the glob.
+        Second element: sorted list of Cassandra directories matched by the glob.
+    """
+    if not _is_valid_glob(graphite_glob):
+        # TODO(d.forest): should we instead raise an exception?
+        return ([], [])
+
+    glob_re = re.compile(_glob_to_regex(graphite_glob))
+    accessor_components = _graphite_glob_to_accessor_components(graphite_glob)
```
Comment (iksaif) [inline biggraphite/glob_utils.py]: I think there still a misunderstanding, in my mind we should have:
* graphite_glob(), used only by the graphite plugin: supports graphite thingies.
* glob(), used by the drivers when the database can't do it: supports star and globstar

Graphite specific things should *stay* in the graphite plugin scope and not leak further.

Comment (dpanth3r) [inline biggraphite/glob_utils.py]: Updated! Added globstar support to glob, too

**Pull request merged**